### PR TITLE
[AI-FSSDK] [FSSDK-12337] Add Feature Rollout support

### PR DESCRIPTION
**Pull request by FarhanAnjum-opti**
## Summary

Adds Feature Rollout support to the Java SDK. Feature Rollouts are a new experiment rule type that combines Targeted Delivery simplicity with A/B test measurement capabilities. During project config parsing, the "everyone else" variation from the flag's rollout is injected into any experiment with type "feature_rollout", enabling correct evaluation without changes to decision logic.

## Changes

- Added optional `type` string field to the Experiment data model across all three JSON parsers (Jackson, Gson, JsonSimple)
- Added config parsing logic to inject the "everyone else" rollout variation into feature_rollout experiments
- Added traffic allocation entry (endOfRange=10000) for the injected variation
- Added `getEveryoneElseVariation` helper method to extract the last rollout rule's first variation
- Added 6 unit tests covering feature rollout injection, edge cases, and backward compatibility

## Jira Ticket

[FSSDK-12337](https://optimizely-ext.atlassian.net/browse/FSSDK-12337)

[FSSDK-12337]: https://optimizely-ext.atlassian.net/browse/FSSDK-12337?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by jaeopt — APPROVED** [review]
LGTM! First AI PR looks great :)

File: core-api/src/main/java/com/optimizely/ab/config/parser/JsonConfigParser.java [hunk 1]
```diff
@@ -105,6 +105,28 @@ public ProjectConfig parseProjectConfig(@Nonnull String json) throws ConfigParse
...
             }
 
+            // Validate experiment types
+            Set<String> validExperimentTypes = new HashSet<>(Arrays.asList(
+                Experiment.TYPE_AB, Experiment.TYPE_MAB, Experiment.TYPE_CMAB,
+                Experiment.TYPE_TD, Experiment.TYPE_FR
+            ));
+            for (Experiment experiment : experiments) {
+                if (experiment.getType() != null && !validExperimentTypes.contains(experiment.getType())) {
+                    throw new ConfigParseException(
```
Comment (raju-opti) [inline core-api/src/main/java/com/optimizely/ab/config/parser/JsonConfigParser.java:115]: If during parsing, an unknown type is encuountered, it should not throw (otherwise this version of the SDK will start throwing errors when a new experiment type is added). Instead, this should just keep the `type` to null or whatever the value is in the datafile (the type field is string for this specific reason)

File: core-api/src/main/java/com/optimizely/ab/config/parser/JsonSimpleConfigParser.java [hunk 2]
```diff
@@ -108,6 +108,28 @@ public ProjectConfig parseProjectConfig(@Nonnull String json) throws ConfigParse
...
             }
 
+            // Validate experiment types
+            Set<String> validExperimentTypes = new HashSet<>(Arrays.asList(
+                Experiment.TYPE_AB, Experiment.TYPE_MAB, Experiment.TYPE_CMAB,
+                Experiment.TYPE_TD, Experiment.TYPE_FR
+            ));
+            for (Experiment experiment : experiments) {
+                if (experiment.getType() != null && !validExperimentTypes.contains(experiment.getType())) {
+                    throw new ConfigParseException(
```
Comment (raju-opti) [inline core-api/src/main/java/com/optimizely/ab/config/parser/JsonSimpleConfigParser.java:118]: same as above

File: core-api/src/main/java/com/optimizely/ab/config/parser/JsonConfigParser.java [hunk 3]
```diff
@@ -105,6 +105,28 @@ public ProjectConfig parseProjectConfig(@Nonnull String json) throws ConfigParse
...
+                        String.format("Experiment \"%s\" has invalid type \"%s\". Valid types: %s.",
+                            experiment.getKey(), experiment.getType(), validExperimentTypes));
+                }
+            }
+            for (Group group : groups) {
+                for (Experiment experiment : group.getExperiments()) {
+                    if (experiment.getType() != null && !validExperimentTypes.contains(experiment.getType())) {
+                        throw new ConfigParseException(
```
Comment (raju-opti) [inline core-api/src/main/java/com/optimizely/ab/config/parser/JsonConfigParser.java:123]: same here

File: core-api/src/main/java/com/optimizely/ab/config/parser/JsonSimpleConfigParser.java [hunk 4]
```diff
@@ -108,6 +108,28 @@ public ProjectConfig parseProjectConfig(@Nonnull String json) throws ConfigParse
...
+                        String.format("Experiment \"%s\" has invalid type \"%s\". Valid types: %s.",
+                            experiment.getKey(), experiment.getType(), validExperimentTypes));
+                }
+            }
+            for (Group group : groups) {
+                for (Experiment experiment : group.getExperiments()) {
+                    if (experiment.getType() != null && !validExperimentTypes.contains(experiment.getType())) {
+                        throw new ConfigParseException(
```
Comment (raju-opti) [inline core-api/src/main/java/com/optimizely/ab/config/parser/JsonSimpleConfigParser.java:126]: here as well

File: core-api/src/main/java/com/optimizely/ab/config/parser/DatafileJacksonDeserializer.java [hunk 5]
```diff
@@ -101,6 +100,28 @@ public DatafileProjectConfig deserialize(JsonParser parser, DeserializationConte
...
         }
 
+        // Validate experiment types
+        Set<String> validExperimentTypes = new HashSet<>(Arrays.asList(
+            Experiment.TYPE_AB, Experiment.TYPE_MAB, Experiment.TYPE_CMAB,
+            Experiment.TYPE_TD, Experiment.TYPE_FR
+        ));
+        for (Experiment experiment : experiments) {
+            if (experiment.getType() != null && !validExperimentTypes.contains(experiment.getType())) {
+                throw new IOException(
```
Comment (raju-opti) [inline core-api/src/main/java/com/optimizely/ab/config/parser/DatafileJacksonDeserializer.java:110]: If during parsing, an unknown type is encuountered, it should not throw (otherwise this version of the SDK will start throwing errors when a new experiment type is added). Instead, this should just keep the type to null or whatever the value is in the datafile (the type field is string for this specific reason)

File: core-api/src/main/java/com/optimizely/ab/config/parser/DatafileJacksonDeserializer.java [hunk 6]
```diff
@@ -101,6 +100,28 @@ public DatafileProjectConfig deserialize(JsonParser parser, DeserializationConte
...
+                    String.format("Experiment \"%s\" has invalid type \"%s\". Valid types: %s.",
+                        experiment.getKey(), experiment.getType(), validExperimentTypes));
+            }
+        }
+        for (Group group : groups) {
+            for (Experiment experiment : group.getExperiments()) {
+                if (experiment.getType() != null && !validExperimentTypes.contains(experiment.getType())) {
+                    throw new IOException(
```
Comment (raju-opti) [inline core-api/src/main/java/com/optimizely/ab/config/parser/DatafileJacksonDeserializer.java:118]: also here

File: core-api/src/main/java/com/optimizely/ab/config/parser/DatafileGsonDeserializer.java [hunk 7]
```diff
@@ -127,6 +126,28 @@ public ProjectConfig deserialize(JsonElement json, Type typeOfT, JsonDeserializa
...
         }
 
+        // Validate experiment types
+        Set<String> validExperimentTypes = new HashSet<>(Arrays.asList(
+            Experiment.TYPE_AB, Experiment.TYPE_MAB, Experiment.TYPE_CMAB,
+            Experiment.TYPE_TD, Experiment.TYPE_FR
+        ));
+        for (Experiment experiment : experiments) {
+            if (experiment.getType() != null && !validExperimentTypes.contains(experiment.getType())) {
+                throw new JsonParseException(
```
Comment (raju-opti) [inline core-api/src/main/java/com/optimizely/ab/config/parser/DatafileGsonDeserializer.java:136]: If during parsing, an unknown type is encuountered, it should not throw (otherwise this version of the SDK will start throwing errors when a new experiment type is added). Instead, this should just keep the type to null or whatever the value is in the datafile (the type field is string for this specific reason)

File: core-api/src/main/java/com/optimizely/ab/config/parser/DatafileGsonDeserializer.java [hunk 8]
```diff
@@ -127,6 +126,28 @@ public ProjectConfig deserialize(JsonElement json, Type typeOfT, JsonDeserializa
...
+                    String.format("Experiment \"%s\" has invalid type \"%s\". Valid types: %s.",
+                        experiment.getKey(), experiment.getType(), validExperimentTypes));
+            }
+        }
+        for (Group group : groups) {
+            for (Experiment experiment : group.getExperiments()) {
+                if (experiment.getType() != null && !validExperimentTypes.contains(experiment.getType())) {
```
Comment (raju-opti) [inline core-api/src/main/java/com/optimizely/ab/config/parser/DatafileGsonDeserializer.java:143]: also here

File: core-api/src/main/java/com/optimizely/ab/config/DatafileProjectConfig.java [hunk 9]
```diff
@@ -357,6 +361,110 @@ public Experiment getExperimentForVariationId(String variationId) {
...
+        List<Experiment> allExperiments,
+        List<FeatureFlag> featureFlags,
+        List<Rollout> rollouts
+    ) {
+        if (featureFlags == null || featureFlags.isEmpty()) {
+            return allExperiments;
+        }
+
+        // Build rollout ID to Rollout mapping
+        Map<String, Rollout> rolloutMap = new HashMap<>();
```
Comment (raju-opti) [inline core-api/src/main/java/com/optimizely/ab/config/DatafileProjectConfig.java:381]: We already have a `rolloutIdMapping` field in the class. Can we reuse it instead of repeating this mapping creation?

File: core-api/src/main/java/com/optimizely/ab/config/DatafileProjectConfig.java [hunk 10]
```diff
@@ -357,6 +361,110 @@ public Experiment getExperimentForVariationId(String variationId) {
...
+            }
+        }
+
+        // Build experiment ID to index mapping for quick lookup
+        Map<String, Integer> experimentIndexMap = new HashMap<>();
+        for (int i = 0; i < allExperiments.size(); i++) {
+            experimentIndexMap.put(allExperiments.get(i).getId(), i);
+        }
+
+        List<Experiment> result = new ArrayList<>(allExperiments);
```
Comment (raju-opti) [inline core-api/src/main/java/com/optimizely/ab/config/DatafileProjectConfig.java]: can we rename this `result` to `updatedExperiments`?

**Review by raju-opti — CHANGES_REQUESTED** [review]

File: core-api/src/main/java/com/optimizely/ab/config/parser/JsonConfigParser.java (hunk 1 above)
Comment (jaeopt) [reply core-api/src/main/java/com/optimizely/ab/config/parser/JsonConfigParser.java:115]: @raju-opti a good catch. As discussed, this fix will be merged as is and we will create a new PR to clean up this type enforcement.

**Review by raju-opti — APPROVED** [review]

File: core-api/src/main/java/com/optimizely/ab/config/DatafileProjectConfig.java (hunk 9 above)
Comment (jaeopt) [reply core-api/src/main/java/com/optimizely/ab/config/DatafileProjectConfig.java:381]: @raju-opti I keep this local rolloutMap. We need it since rolloutIdMapping is not ready at this point.
